clash-lib/src/Clash/Normalize/Strategy.hs

```diff
@@ -26,7 +26,7 @@ normalization = rmDeadcode >-> constantPropgation >-> etaTL >-> rmUnusedExpr >-!
     rmUnusedExpr = bottomupR (apply "removeUnusedExpr" removeUnusedExpr)
     rmDeadcode = bottomupR (apply "deadcode" deadCode)
     bindConst  = topdownR (apply "bindConstantVar" bindConstantVar)
-    evalConst  = topdownR (apply "evalConst" reduceConst)
+    evalConst  = bottomupR (apply "evalConst" reduceConst)
```

Comment:
Add the reason for this bottomup traversal as a comment/note `-- See [Note] bottomup traversal evalConst`.

Also, I don't completely agree with the nuance of the commit message. Basically, there are two issues:

1. 2-May-2019: There is a bug in the evaluator where all data constructors are considered lazy, even though their declaration says they have strict fields. This causes some reductions to fail because the term under the constructor is not in WHNF, which is what some of the evaluation rules for certain primitive operations expect. Using a bottom-up traversal works around this bug by ensuring that the values under the constructor are in WHNF
2. Using a bottomup traversal ensures that constants are reduced to NF, even if constructors are lazy, thus ensuring more sensible/smaller generated HDL.

Reply:
I see. I should have explicitly stated that the top-down traversal in and of itself is not a problem and that it only becomes a problem if you don't account for strictness annotations. Thanks!

Reply:
Just so we're clear, these are the only changes you'd like to see? You're fine with merging even with the issues I outlined in the OP?

Reply:
Yes, we can fix them later.